format/ComputeIR.fbs

```diff
@@ -0,0 +1,241 @@
+  /// 0 refers to the left relation and 1 to the right relation.
+  relation_index: int;
+
+  /// The type of data in the referenced Field.
+  type: Type;
+}
+
+table Call {
+  /// The name of the function whose invocation this Call represents.
+  function_name: string (required);
+
+  /// Parameters for `function_name`; content/format may be unique to each
+  /// value of `function_name`.
+  options: Buffer;
+
+  /// The arguments passed to `function_name`.
+  arguments: [Expression] (required);
+
+  /// The type of data which invoking `function_name` will return.
+  type: Type;
+}
+
+/// A relation is a set of rows with consistent schema.
+table Relation {
+  /// The namespaced name of this Relation.
+  ///
+  /// Names with no namespace are reserved for pure relational
+  /// algebraic operations, which currently include:
+  ///   "filter"
+  ///   "project"
+  ///   "aggregate"
+  ///   "join"
+  ///   "order_by"
+  ///   "limit"
+  ///   "common"
+  ///   "union"
+  ///   "literal"
+  ///   "interactive_output"
+  relation_name: string (required);
+
+  /// Parameters for `relation_name`; content/format may be unique to each
+  /// value of `relation_name`.
+  options: Buffer;
+
+  /// The arguments passed to `relation_name`.
+  arguments: [Relation] (required);
+
+  /// The schema of rows in this Relation
+  schema: Schema;
+}
+
+/// The contents of Relation.options will be FilterOptions
+/// if Relation.relation_name = "filter"
+table FilterOptions {
+  /// The expression which will be evaluated against input rows
+  /// to determine whether they should be excluded from the
+  /// "filter" relation's output.
+  filter_expression: Expression (required);
+}
+
+/// The contents of Relation.options will be ProjectOptions
+/// if Relation.relation_name = "project"
+table ProjectOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "project" relation's output.
+  expressions: [Expression] (required);
+}
+
+/// The contents of Relation.options will be AggregateOptions
+/// if Relation.relation_name = "aggregate"
+table AggregateOptions {
+  /// Expressions which will be evaluated to produce to
+  /// the rows of the "aggregate" relation's output.
+  aggregations: [Expression] (required);
+  /// Keys by which `aggregations` will be grouped.
+  keys: [Expression];
+}
+
+/// The contents of Relation.options will be JoinOptions
+/// if Relation.relation_name = "join"
+table JoinOptions {
+  /// The expression which will be evaluated against rows from each
+  /// input to determine whether they should be included in the
+  /// "join" relation's output.
+  on_expression: Expression (required);
+  join_kind: string;
+}
+
+/// Whether lesser values should precede greater or vice versa,
+/// also whether nulls should preced or follow values.
+enum Ordering : uint8 {
+  ASCENDING_THEN_NULLS,
+  DESCENDING_THEN_NULLS,
+  NULLS_THEN_ASCENDING,
+  NULLS_THEN_DESCENDING
```

Comment:
What was the rationale for combining these into one enum?

Reply:
It seems clearer as a single enum, based on discussion https://github.com/apache/arrow/pull/10934#discussion_r688802634

Reply:
Cool, a producer can always orthogonalize the API if it so desires.